src/include/libpmemobj.h

```diff
@@ -84,6 +99,505 @@ void pmemobj_set_funcs(
+ * be obtained by calling pmemobj_tx_get_stage.
+ *
+ * If successful, function returns zero. Otherwise, an error number is returned.
+ */
+int pmemobj_tx_process();
+
+#define	__TX_BEGIN(pop, locks, ...)					\
+{									\
+	jmp_buf __env;							\
+	pmemobj_tx_begin(pop, &__env, locks, ##__VA_ARGS__);		\
+	setjmp(__env);							\
+	switch ((int)pmemobj_tx_get_stage()) {				\
+		case TX_STAGE_WORK:					\
+
+#define	TX_BEGIN_WRLOCK(pop, ...) __TX_BEGIN(pop, 1, ##__VA_ARGS__)
+
+#define	TX_BEGIN_LOCK(pop, ...) __TX_BEGIN(pop, 0, ##__VA_ARGS__)
+
+#define	TX_BEGIN(pop) TX_BEGIN_LOCK(pop)
+
+#define	TX_ONABORT							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_ONABORT:
+
+#define	TX_ONCOMMIT							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_ONCOMMIT:
+
+/* called regardless of the transaction end state */
+#define	TX_FINALLY							\
+			pmemobj_tx_process();				\
+			break;						\
+		case TX_STAGE_FINALLY:
+
+#define	TX_END								\
+			pmemobj_tx_process();				\
+			break;						\
+		default:						\
+			pmemobj_tx_process();				\
+			break;						\
+	}								\
+	pmemobj_tx_end();						\
+}
+
+/*
+ * Takes a "snapshot" of the memory block of given size and located at given
+ * offset, and saves it in the undo log.
+ * The application is then free to directly modify the object in that memory
+ * range. In case of failure or abort, all the changes within this range will
+ * be rolled-back automatically.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+int pmemobj_tx_add_range(uint64_t off, size_t size);
+
+/*
+ * Transactionally allocates new object.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_alloc(size_t size, int type_num);
+
+/*
+ * Transactionally allocates new zeroed object.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_zalloc(size_t size, int type_num);
+
+/*
+ * Transactionally resizes an existing object.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+PMEMoid pmemobj_tx_realloc(PMEMoid oid, size_t size, int type_num);
+
+/*
+ * Transactionally frees existing object.
+ *
+ * If successful and called during TX_STAGE_WORK, function returns zero.
+ * Otherwise, state changes to TX_STAGE_ONABORT and an error number is returned.
+ */
+int pmemobj_tx_free(PMEMoid oid);
+
+#define	TX_ADD_RANGE(start_offset, size)				\
+pmemobj_tx_add_range(start_offset, size);
+
+#define	TX_ADD(o)							\
+TX_ADD_RANGE(o.oid.off, sizeof (*o.__type));
```

Comment:
It's usually a good idea to use () around argument names in the macro body.
`(o).oid.off`, `*(o).__type`